e2etest/bgptests/bgp.go

```diff
@@ -1209,6 +1209,185 @@ var _ = ginkgo.Describe("BGP", func() {
+					Pools: pools,
+					Peers: metallb.PeersForContainers(FRRContainers, ipFamily),
+					BGPAdvs: []metallbv1beta1.BGPAdvertisement{
+						{
+							ObjectMeta: metav1.ObjectMeta{
+								Name: "test-advertisement1",
+							},
+						},
+					},
+				}
+				err = ConfigUpdater.Update(resources)
+				framework.ExpectNoError(err)
+
+				ginkgo.By(fmt.Sprintf("configure service number %d", i+1))
+				svc, _ := testservice.CreateWithBackend(cs, f.Namespace.Name, fmt.Sprintf("svc%d", i+1), testservice.TrafficPolicyCluster, func(svc *corev1.Service) {
+					svc.Annotations = map[string]string{"metallb.universe.tf/address-pool": fmt.Sprintf("test-addresspool%d", i+1)}
+				})
+				defer testservice.Delete(cs, svc)
+
+				for _, c := range FRRContainers {
+					ginkgo.By(fmt.Sprintf("checking connectivity of service %d to external frr container %s", i+1, c.Name))
+					validateService(cs, svc, allNodes.Items, c)
+				}
+
+				services = append(services, svc)
+			}
+
+			var bgpAdvs []metallbv1beta1.BGPAdvertisement
+			for i := 0; i < 2; i++ {
+				ginkgo.By(fmt.Sprintf("setting peer selector for addresspool number %d to peer %s", i+1, testContainers[i].Name))
+				peer := ""
+				for _, p := range resources.Peers {
+					if strings.Contains(p.Name, testContainers[i].Name) {
+						peer = p.Name
+					}
+				}
+				bgpAdv := metallbv1beta1.BGPAdvertisement{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: fmt.Sprintf("test-advertisement%d", i+1),
+					},
+					Spec: metallbv1beta1.BGPAdvertisementSpec{
+						IPAddressPools: []string{fmt.Sprintf("test-addresspool%d", i+1)},
+						Peers:          []string{peer},
+					},
+				}
+				bgpAdvs = append(bgpAdvs, bgpAdv)
+
+				resources := metallbconfig.ClusterResources{
+					Pools:   pools,
+					Peers:   metallb.PeersForContainers(FRRContainers, ipFamily),
+					BGPAdvs: bgpAdvs,
+				}
+				err = ConfigUpdater.Update(resources)
+				framework.ExpectNoError(err)
+
+				for _, c := range FRRContainers {
+					if testContainers[i].Name == c.Name {
+						ginkgo.By(fmt.Sprintf("checking connectivity of service %d to external frr container %s", i+1, c.Name))
+						validateService(cs, services[i], allNodes.Items, c)
+					} else {
+						ginkgo.By(fmt.Sprintf("checking service %d not advertised to external frr container %s", i+1, c.Name))
+						validateServiceNotAdvertised(services[i], c, ipFamily)
+					}
+				}
+
+				ginkgo.By(fmt.Sprintf("removing peer selector from bgpadvertisement %d", i+1))
+				resources.BGPAdvs[i].Spec.Peers = nil
+				err = ConfigUpdater.Update(resources)
+				framework.ExpectNoError(err)
+
+				for _, c := range FRRContainers {
+					ginkgo.By(fmt.Sprintf("checking connectivity of service %d to external frr container %s", i+1, c.Name))
+					validateService(cs, services[i], allNodes.Items, c)
+				}
+			}
+
+			ginkgo.By("checking same service advertised through two different advertisements to two different peers")
+			for i := 0; i < 2; i++ {
```

Comment:
Same considerations as above, following by index is hard. Be explicit on naming.
One pool, PeerForAdv1 , PeerForAdv2, etc

Also, this should be a different test and not a by in the same. It's already long.